src/sm.c

```diff
@@ -0,0 +1,480 @@
+    {
+        LogError("sm name=%s. cannot sm_close_end_internal that which is in %" PRI_SM_STATE " state", sm->name, SM_STATE_VALUE(state));
+    }
+    else
+    {
+        /*Codes_SRS_SM_02_044: [ sm_close_end shall switch the state to SM_CREATED. ]*/
+        if (InterlockedCompareExchange(&sm->state, state - SM_CLOSING + SM_CREATED + SM_STATE_INCREMENT, state) != state)
+        {
+            LogError("sm name=%s. state changed meanwhile (it was %" PRI_SM_STATE "), likely competing threads", sm->name, SM_STATE_VALUE(state));
+        }
+        else
+        {
+
+        }
+    }
+}
+
+void sm_close_end(SM_HANDLE sm)
+{
+    /*Codes_SRS_SM_02_018: [ If sm is NULL then sm_close_end shall return. ]*/
+    if (sm == NULL)
+    {
+        LogError("invalid argument SM_HANDLE sm=%p", sm);
+    }
+    else
+    {
+        sm_close_end_internal(sm);
+    }
+}
+
+SM_RESULT sm_exec_begin(SM_HANDLE sm)
+{
+    SM_RESULT result;
+    /*Codes_SRS_SM_02_021: [ If sm is NULL then sm_exec_begin shall fail and return SM_ERROR. ]*/
+    if (sm == NULL)
+    {
+        LogError("invalid argument SM_HANDLE sm=%p", sm);
+        result = SM_ERROR;
+    }
+    else
+    {
+        LONG state1 = InterlockedAdd(&sm->state, 0);
+        if (
+            /*Codes_SRS_SM_02_054: [ If state is not SM_OPENED then sm_exec_begin shall return SM_EXEC_REFUSED. ]*/
+            ((state1 & SM_STATE_MASK) != SM_OPENED) ||
+            /*Codes_SRS_SM_02_055: [ If SM_CLOSE_BIT is 1 then sm_exec_begin shall return SM_EXEC_REFUSED. ]*/
+            ((state1 & SM_CLOSE_BIT) == SM_CLOSE_BIT)
+            )
+        {
+            LogError("sm name=%s. cannot call sm_exec_begin when state is %" PRI_SM_STATE "", sm->name, SM_STATE_VALUE(state1));
+            result = SM_EXEC_REFUSED;
+        }
+        else
+        {
+            /*Codes_SRS_SM_02_056: [ sm_exec_begin shall increment n. ]*/
+            (void)InterlockedIncrement(&sm->n);
+            LONG state2 = InterlockedAdd(&sm->state, 0);
+
+            /*Codes_SRS_SM_02_057: [ If the state changed after incrementing n then sm_exec_begin shall return SM_EXEC_REFUSED. ]*/
+            if (state1 != state2)
+            {
+                LogError("sm name=%s. state changed meanwhile from %" PRI_SM_STATE " to %" PRI_SM_STATE "", sm->name, SM_STATE_VALUE(state1), SM_STATE_VALUE(state2));
+                LONG n = InterlockedDecrement(&sm->n);
+                if (n == 0)
+                {
+                    WakeByAddressSingle((void*)&sm->n);
+                }
+                result = SM_EXEC_REFUSED;
+            }
+            else
+            {
+                /*Codes_SRS_SM_02_058: [ sm_exec_begin shall return SM_EXEC_GRANTED. ]*/
+                result = SM_EXEC_GRANTED;
+            }
+        }
+    }
+    return result;
+}
+
+void sm_exec_end(SM_HANDLE sm)
+{
+    /*Codes_SRS_SM_02_024: [ If sm is NULL then sm_exec_end shall return. ]*/
+    if (sm == NULL)
+    {
+        LogError("return");
```

Comment:
>return [](start = 18, length = 6)

Maybe "NULL sm"

Reply:
Fixed

---
In reply to: [423416343](https://github.com/Azure/azure-c-util/pull/37#discussion_r423416343) [](ancestors = 423416343)